pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,337 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package populators
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"kubevirt.io/containerized-data-importer/pkg/common"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	featuregates "kubevirt.io/containerized-data-importer/pkg/feature-gates"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type pvcModifierFunc func(pvc *corev1.PersistentVolumeClaim, source client.Object)
+
+// ReconcilerBase members
+type ReconcilerBase struct {
+	client          client.Client
+	recorder        record.EventRecorder
+	scheme          *runtime.Scheme
+	log             logr.Logger
+	featureGates    featuregates.FeatureGates
+	sourceKind      string
+	installerLabels map[string]string
+}
+
+const (
+	dataSourceRefField = "spec.dataSourceRef"
+)
+
+// Interface to store populator-specific methods
+type populatorController interface {
+	// Returns the specific populator CR
+	getPopulationSource(namespace, name string) (client.Object, error)
+	// Returns PVC' annotations according to the population source
+	updateAnnotations(pvc *corev1.PersistentVolumeClaim, source client.Object)
+	// Reconciles the target PVC with populator-specific logic
+	reconcileTargetPVC(pvc, pvcPrime *corev1.PersistentVolumeClaim) (reconcile.Result, error)
+}
+
+// CreateCommonPopulatorIndexes creates indexes used by all populators
+func CreateCommonPopulatorIndexes(mgr manager.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &corev1.PersistentVolumeClaim{}, dataSourceRefField, func(obj client.Object) []string {
+		pvc := obj.(*corev1.PersistentVolumeClaim)
+		dataSourceRef := pvc.Spec.DataSourceRef
+		if dataSourceRef != nil && dataSourceRef.APIGroup != nil &&
+			*dataSourceRef.APIGroup == cc.AnnAPIGroup && dataSourceRef.Name != "" {
+			return []string{getPopulatorIndexKey(obj.GetNamespace(), pvc.Spec.DataSourceRef.Kind, pvc.Spec.DataSourceRef.Name)}
```

Comment:
since rebased on 1.26 should check `pvc.Spec.DataSourceRef.Namespace)` before using `pvc.Namespace`

Reply:
Planned to work on cross-namespace population in a separate PR but yeah I'll do it now.

Reply:
Just committed some changes related to cross-namespace population. I've not been able to test them yet since only k8s v1.26 allows `CrossNamespaceVolumeDataSource` (kubevirtci still uses v1.25.7).

Another problem that I see is that there's no way to check if `CrossNamespaceVolumeDataSource` is enabled (still alpha), and there are more [prerequisites](https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/#trying-it-out) than I expected, so we should probably just skip that test for the moment.